File: servers/go/internal/application/queries/lending.go
package queries

import (
	"context"
	"fmt"

	"github.com/christophercaldwell/model-architecture/go/internal/domain/loan"
	"github.com/christophercaldwell/model-architecture/go/internal/domain/member"
)

type LendingQueries struct {
	loanRepo loan.ReadRepository
}

func NewLendingQueries(loanRepo loan.ReadRepository) *LendingQueries {
	return &LendingQueries{loanRepo: loanRepo}
}

func (q *LendingQueries) GetMemberLoans(ctx context.Context, ident member.MemberIdent) ([]loan.Loan, error) {
	loans, err := q.loanRepo.GetByMemberIdent(ctx, ident)
	if err != nil {
		return nil, fmt.Errorf("get member loans: %w", err)
	}
	return loans, nil
}

func (q *LendingQueries) GetOverdueLoans(ctx context.Context) ([]loan.Loan, error) {
	loans, err := q.loanRepo.GetOverdue(ctx)
	if err != nil {
		return nil, fmt.Errorf("get overdue loans: %w", err)
	}
	return loans, nil
}

File: server/crates/application/src/queries/lending.rs
use domain::{
    loan::{port::LoanReadRepoPort, Loan},
    member::MemberIdent,
};
use std::sync::Arc;

use super::QueryError;

#[derive(Clone)]
pub struct LendingQueries {
    loan_read_repo: Arc<dyn LoanReadRepoPort>,
}

impl LendingQueries {
    #[must_use]
    pub fn new(loan_read_repo: Arc<dyn LoanReadRepoPort>) -> Self {
        Self { loan_read_repo }
    }

    pub async fn get_member_loans(&self, ident: &MemberIdent) -> Result<Vec<Loan>, QueryError> {
        self.loan_read_repo
            .get_by_member_ident(ident)
            .await
            .map_err(QueryError::from)
    }

    pub async fn get_overdue_loans(&self) -> Result<Vec<Loan>, QueryError> {
        self.loan_read_repo
            .get_overdue()
            .await
            .map_err(QueryError::from)
    }
}
